File: internal/http/dto/discovery.go
package dto

import (
	"time"

	"salespilot/internal/domain"
)

// --- Request ---

// DiscoveryRunRequest is the optional body for POST /api/discovery/run.
// CorrelationKey lets a caller (e.g. the scheduler, TK-12.5.1) make a
// trigger idempotent; omitted/empty always starts a fresh run.
type DiscoveryRunRequest struct {
	CorrelationKey *string `json:"correlation_key" validate:"omitempty"`
}

// --- Response ---

type DiscoveryRunResponse struct {
	ID             string     `json:"id"`
	StartedAt      time.Time  `json:"started_at"`
	FinishedAt     *time.Time `json:"finished_at"`
	SourceIDs      []string   `json:"source_ids"`
	Status         string     `json:"status"`
	FoundCount     int        `json:"found_count"`
	Summary        *string    `json:"summary"`
	CorrelationKey *string    `json:"correlation_key"`
	CreatedAt      time.Time  `json:"created_at"`
	UpdatedAt      time.Time  `json:"updated_at"`
}

type DiscoveryRunListResponse struct {
	Items    []DiscoveryRunResponse `json:"items"`
	Total    int64                  `json:"total"`
	Page     int                    `json:"page"`
	PageSize int                    `json:"page_size"`
}

// ToDiscoveryRunResponse maps domain.DiscoveryRun → DiscoveryRunResponse.
func ToDiscoveryRunResponse(r domain.DiscoveryRun) DiscoveryRunResponse {
	sourceIDs := r.SourceIDs
	if sourceIDs == nil {
		sourceIDs = []string{}
	}
	return DiscoveryRunResponse{
		ID:             r.ID,
		StartedAt:      r.StartedAt,
		FinishedAt:     r.FinishedAt,
		SourceIDs:      sourceIDs,
		Status:         string(r.Status),
		FoundCount:     r.FoundCount,
		Summary:        r.Summary,
		CorrelationKey: r.CorrelationKey,
		CreatedAt:      r.CreatedAt,
		UpdatedAt:      r.UpdatedAt,
	}
}

File: apps/web/src/api/discovery.ts
import { useQuery, useMutation, useQueryClient } from '@tanstack/react-query'
import { apiFetch, buildQueryString } from '../lib/api'
import type { TenderApiAction, TenderListResponse } from './tenders'

// ── Types ─────────────────────────────────────────────────────────────────────

export type DiscoveryRunStatus = 'pending' | 'running' | 'success' | 'failed'

export interface DiscoveryRun {
  id: string
  started_at: string
  finished_at: string | null
  source_ids: string[]
  status: DiscoveryRunStatus
  found_count: number
  summary: string | null
  correlation_key: string | null
  created_at: string
  updated_at: string
}

export interface DiscoveryRunListResponse {
  items: DiscoveryRun[]
  total: number
  page: number
  page_size: number
}

export interface DiscoveryInboxFilters {
  recommended_action?: TenderApiAction
  min_score?: number
  page?: number
  page_size?: number
}

// ── Query Hooks ───────────────────────────────────────────────────────────────

export function useDiscoveryInbox(filters: DiscoveryInboxFilters = {}) {
  return useQuery({
    queryKey: ['discovery-inbox', filters],
    queryFn: () => apiFetch<TenderListResponse>(`/api/discovery/inbox${buildQueryString({ ...filters })}`),
  })
}

/** Recent runs, newest first — used for header status + crawl-in-progress detection. */
export function useDiscoveryRuns(options: { refetchInterval?: number | false } = {}) {
  return useQuery({
    queryKey: ['discovery-runs'],
    queryFn: () => apiFetch<DiscoveryRunListResponse>('/api/discovery/runs?page_size=5'),
    refetchInterval: options.refetchInterval,
  })
}

// ── Mutation Hooks ────────────────────────────────────────────────────────────

export function useRunDiscovery() {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: () => apiFetch<DiscoveryRun>('/api/discovery/run', { method: 'POST' }),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['discovery-runs'] })
      queryClient.invalidateQueries({ queryKey: ['discovery-inbox'] })
    },
  })
}
